sys/dev/isp/isp.c

```diff
@@ -4322,18 +4340,362 @@ isp_reinit(ispsoftc_t *isp, int do_load_defaults)
+		} else if (IS_27XX(isp)) {
+			fcp->flash_data_addr = ISP27XX_BASE_ADDR;
+			fcp->flt_region_flt = ISP27XX_FLT_ADDR;
+		} else if (IS_25XX(isp)) {
+			fcp->flash_data_addr = ISP25XX_BASE_ADDR;
+			fcp->flt_region_flt = ISP25XX_FLT_ADDR;
+		} else {
+			fcp->flash_data_addr = ISP24XX_BASE_ADDR;
+			fcp->flt_region_flt = ISP24XX_FLT_ADDR;
+		}
+		fcp->flt_length = 0;
+		r = isp_read_flthdr_2xxx(isp);
+		if (r == 0) {
+			isp_read_flt_2xxx(isp);
+		} else { /* fallback to hardcoded NVRAM address */
+			if (IS_28XX(isp)) {
+				fcp->flt_region_nvram = 0x300000;
+			} else if (IS_27XX(isp)) {
+				fcp->flash_data_addr = 0x7fe7c000;
+				fcp->flt_region_nvram = 0;
+			} else if (IS_25XX(isp)) {
+				fcp->flt_region_nvram = 0x48000;
+			} else {
+				fcp->flash_data_addr = 0x7ffe0000;
+				fcp->flt_region_nvram = 0;
+			}
+			fcp->flt_region_nvram += ISP2400_NVRAM_PORT_ADDR(isp->isp_port);
+		}
+	} else {
+		fcp->flash_data_addr = 0x7fe7c000;
+		fcp->flt_region_nvram = 0;
+		fcp->flt_region_nvram += ISP2400_NVRAM_PORT_ADDR(isp->isp_port);
+	}
 	return (isp_read_nvram_2400(isp));
 }
 
+static int
+isp_read_flthdr_2xxx(ispsoftc_t *isp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int retval = 0;
+	uint32_t addr, lwrds, *dptr;
+	uint16_t csum;
+	uint8_t flthdr_data[FLT_HEADER_SIZE];
+
+	addr = fcp->flt_region_flt;
+	dptr = (uint32_t *) flthdr_data;
+
+	isp_prt(isp, ISP_LOGDEBUG0,
+	    "FLTL[DEF]: 0x%x", addr);
+	for (lwrds = 0; lwrds < FLT_HEADER_SIZE >> 2; lwrds++) {
+		isp_rd_2xxx_flthdr(isp, addr++, dptr++);
+	}
+	dptr = (uint32_t *) flthdr_data;
+	for (csum = 0, lwrds = 0; lwrds < FLT_HEADER_SIZE >> 4; lwrds++) {
+		uint16_t tmp;
+		ISP_IOXGET_16(isp, &dptr[lwrds], tmp);
+		csum += tmp;
+	}
+	if (csum != 0) {
+		retval = -1;
+		goto out;
+	}
+	isp_parse_flthdr_2xxx(isp, flthdr_data);
+out:
+	return (retval);
+}
+
+static void
+isp_rd_2xxx_flthdr(ispsoftc_t *isp, uint32_t addr, uint32_t *rp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int loops = 0;
+	uint32_t base = fcp->flash_data_addr;
+	uint32_t tmp = 0;
+
+	ISP_WRITE(isp, BIU2400_FLASH_ADDR, base | addr);
+	for (loops = 0; loops < 5000; loops++) {
+		ISP_DELAY(10);
+		tmp = ISP_READ(isp, BIU2400_FLASH_ADDR);
+		if ((tmp & (1U << 31)) != 0) {
+			break;
+		}
+	}
+	if (tmp & (1U << 31)) {
+		*rp = ISP_READ(isp, BIU2400_FLASH_DATA);
+		ISP_SWIZZLE_NVRAM_LONG(isp, rp);
+	} else {
+		*rp = 0xffffffff;
+	}
+}
+
+static void
+isp_parse_flthdr_2xxx(ispsoftc_t *isp, uint8_t *flthdr_data)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	uint16_t ver, csum;
+
+	ver = le16toh((uint16_t) (ISP2XXX_FLT_VERSION(flthdr_data)));
+	fcp->flt_length = le16toh((uint16_t) (ISP2XXX_FLT_LENGTH(flthdr_data)));
+	csum = le16toh((uint16_t) (ISP2XXX_FLT_CSUM(flthdr_data)));
+
+	if ((fcp->flt_length == 0 ) ||
+	    (fcp->flt_length > (FLT_HEADER_SIZE + FLT_REGIONS_SIZE))) {
+		isp_prt(isp, ISP_LOGERR,
+		    "FLT[DEF]: Invalid length=0x%x(%d)",
+		    fcp->flt_length, fcp->flt_length);
+	}
+	isp_prt(isp, ISP_LOGCONFIG,
```

Comment:
ISP_LOGCONFIG is enabled by default.  I am not sure internal data structures of the flash are useful to average user.  I feel they should be on some debug level.